### DISCUSSION TITLE
How to write toga dummy backend code

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#2064] [Q&A]
In the dummy backend, we need to write the same backend functions as on other backends. So, how does one write a:

* Function that returns a value:
&nbsp;&nbsp;&nbsp;&nbsp;Directly return the value with `return`?
* Function that returns an object:
&nbsp;&nbsp;&nbsp;&nbsp;I have no idea, how to do this.
* Function that gets the value from an attribute:
&nbsp;&nbsp;&nbsp;&nbsp;Using `self._get_value()`?
* Function that sets the value to an attribute:
&nbsp;&nbsp;&nbsp;&nbsp;Using `self._set_value()`?
* Function that does not return anything:
&nbsp;&nbsp;&nbsp;&nbsp;Using `self._action()`? 

Also it would be really helpful, if someone could provide an example usage of accomplishing these tasks in the dummy backend.

**Comment by freakboy3742** [comment]
I'm a little confused about these questions. The dummy backend isn't some magical special language - it's Python code. So - how do you write a function that returns a value? Like this:

```
def some_function(...):
    return some_value
```

The *behavior* of the dummy backend is design to be the same as a "normal" GUI backend, but without *actually* creating a GUI or having a dependency on a GUI framework. It's a mock - a simulation of what a "real" backend would do. To that end:

 * If the backend has persistent state, use `_set_value()` to save that state, and `_get_value()` to retrieve that state
 * If the backend does some action that doesn't have state, use `_action` to record that the action occurred.

As an example - when you give a widget focus, that's an action - so `focus()` invokes `_action()`. When you set a widget as enabled, that's stateful (because you might want to query the enabled/disabled state of the widget later), so you use `_set_value()` and `_get_value()`.

As for examples... the entire dummy backend is full of examples. 

If you need help with a specific task, you'll need to ask specific question.

**Reply by proneon267** [reply]
Yes, I need help with a specific task:
I have a backend function called `get_primary_screen()` that returns a reference to the primary screen as `screen._impl`. How should I simulate this in the dummy backend?

**Reply by freakboy3742** [reply]
By pretending. 

Dummy doesn't *actually* have screens - But it needs to *pretend* as if it actually has screens. Provide a dummy implementation of Screen as if it were a normal GUI backend like Cocoa or GTK; but instead of calling a native API to retrieve details about a specific, real screen, or to change properties of an actual screen - fake it, and return something plausible, or log the fact that you did something with an `_action`. Pretend that the dummy has 2 screens (that should be enough for testing purposes); and pick one of them as the primary.

**Reply by proneon267** [reply]
Thanks. I will try this out and let you know.
